CI.rst

```diff
@@ -629,6 +628,90 @@ delete old artifacts that are > 7 days old. It only runs for the 'apache/airflow
+   files from that commit directly).
+
+3) If any of the important, environment files changed (Dockerfile, ci scripts, setup.py, github workflow
+   files), then we again run all tests and checks. Those are cases where the logic of the checks changed
+   or the environment for the checks changed so we want to make sure to check everything.
+
+4) If any of docs changed: we need to have CI image so we enable image building
+
+5) If any of chart files changed, we need to run helm tests so we enable helm unit tests
+
+6) If any of API files changed, we need to run API tests so we enable them
+
+7) if any of the relevant source files that trigger the tests have changed at all. Those are airflow
+   sources, chart, tests and kubernetes_tests. If any of those files changed, we enable tests and we
+   enable image building, because the CI images are needed to run tests.
+
+8) Then we determine which types of the tests should be run. We count all the changed files in the
+   relevant airflow sources (airflow, chart, tests, kubernetes_tests) first and then we count how many
+   files changed in different packages:
+
+   a) if any of the API files changed we enable ``API`` test type
+   b) if any of the CLI files changed we enable ``CLI`` test type
+   c) if any of the Provider files changed we enable ``Providers`` test type
+   d) if any of the WWW files changed we enable ``WWW`` test type
+   e) if any of the Kubernetes files changed we enable ``Kubernetes`` test type
+   f) Then we subtract count of all the ``specific`` above per-type changed files from the count of
+      all changed files. In case there are any files changed, then we assume that some unknown files
+      changed (likely from the core of airflow) and in this case we enable all test types above and the
+      Core test types - simply because we do not want to risk to miss anything.
+    g) In all cases where tests are enabled we also add Heisentests, Integration and - depending on
+       the backend used = Postgres or MySQL types of tests.
+
+9) Quarantined tests are always run when tests are run: - we need to run them often to observe how
+   often they fail so that we can decide to move them out of quarantine. Details about the
+   Quarantined tests are described in `TESTING.rst <TESTING.rst>`_
+
+10) There is a special case of static checks. In case the above logic determines that teh CI image
+    needs to be build, we run long and more comprehensive version of static checks - including Pylint,
+    MyPy, Flake8. And those tests are run on all files, no matter how many files changed.
+    In case the image is not built, we run only simpler set of changes - the longer static checks
+    that require CI image are skipped, and we only run the tests on the files that changed in the incoming
+    commit - unlike pylint/flake8/mypy, those static checks are per-file based and they should not miss any
+    important change.
```

Comment:
Good explanation, well done @potiuk 👏

Reply:
Ahh.. :heart: .. It was REALLY difficult to explain it :)